core/chain_manager.go

```diff
@@ -525,10 +529,27 @@ func (self *ChainManager) InsertChain(chain types.Blocks) (int, error) {
 		stats      struct{ queued, processed, ignored int }
 		tstart     = time.Now()
 	)
+
+	// check the nonce in parallel to the block processing
+	// this speeds catching up significantly
+	errch := make(chan error, len(chain))
+	go func() {
+		for _, block := range chain {
+			go func(block *types.Block) {
+				if !self.pow.Verify(block) {
+					errch <- ValidationError("Block's nonce is invalid (= %x)", block.Nonce)
+				} else {
+					errch <- nil
+				}
+			}(block)
+		}
+	}()
+
```

Comment:
Depending on the length of `chain`, this could generate more computation than the processor can handle.
Might be a good idea to schedule this onto a fixed number (`GOMAXPROCS`) of workers.

```go
func verifyNonces(blocks []*types.Block) error {
	// Spawn a few workers. They listen for blocks on the in channel
	// and send results on done. The workers will exit in the
	// background when in is closed.
	var (
		in   = make(chan *types.Block)
		done = make(chan error, runtime.GOMAXPROCS(0))
	)
	defer close(in)
	for i := 0; i < cap(done); i++ {
		go verifyNonce(in, done)
	}
	// Feed blocks to the workers, aborting at the first invalid nonce.
	var (
		running, i int
		block      *types.Block
		sendin     = in
	)
	for i < len(blocks) || running > 0 {
		if i == len(blocks) {
			// Disable sending to in.
			sendin = nil
		} else {
			block = blocks[i]
			i++
		}
		select {
		case sendin <- block:
			running++
		case err := <-done:
			running--
			if err != nil {
				return err
			}
		}
	}
	return nil
}

func verifyNonce(pow pow.Pow, in <-chan *types.Block, done chan<- error) {
	for block := range in {
		if !pow.Verify(block) {
			done <- ValidationError(...)
		} else {
			done <- nil
		}
	}
}
```